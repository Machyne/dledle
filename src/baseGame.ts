// Give these string values so they can be round trip through being used as object keys.
// These are never encoded as-is so strings are fine.
export enum GameScore {
  Win = "Win",
  NearWin = "NearWin",
  Loss = "Loss",
}

export type GameResult = {
  score: GameScore;
  serializedResult: string;
};

export abstract class BaseGame {
  private _resultRegex: RegExp | null = null;

  /**
   * The name of the game.
   */
  abstract get name(): string;

  /**
   * The name of the game in lowercase with only alphanumeric characters.
   */
  get fileName(): string {
    return this.name.toLowerCase().replace(/[^a-z0-9]/g, "");
  }

  /**
   * A regex that matches the result of the game.
   */
  get resultRegex(): RegExp {
    if (this._resultRegex === null) {
      this._resultRegex = this._buildResultRegex();
    }
    return this._resultRegex;
  }

  /**
   * Builds a regex that matches the result of the game. The return value will be cached.
   */
  abstract _buildResultRegex(): RegExp;

  /**
   * A link to the game.
   */
  abstract get link(): string;

  /**
   * Parses the user-supplied game result and returns a serialized version of
   * the game data in addition to the "score".
   *
   * @param gameResult the result of the game run through the resultRegex
   */
  abstract serializeResult(gameResult: RegExpMatchArray): GameResult;

  /**
   * The share text of the result of the game given the serialized game data.
   *
   * @param serializedResult the serialized game data
   */
  abstract deserialize(serializedResult: string): string;
}
