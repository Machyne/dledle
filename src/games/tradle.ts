import { BaseGame, GameScore } from "../baseGame";
import { DEFAULT_WORD_WIDTH, intToBase64, nextInt } from "../util/base64";
import { emojiSquares, fiveSquaresRegex, splitEmojiLines } from "../util/emoji";
import { emojiSquareScaleToNumber, numberToEmojiSquareScale } from "../util/gameHelpers";

const guessEncodingWidth = 2;

export class Tradle extends BaseGame {
  get name(): string {
    return "Tradle";
  }

  get link(): string {
    return "https://oec.world/en/tradle";
  }

  _buildResultRegex(): RegExp {
    return new RegExp(`#Tradle #(\\d+) ([1-6X])\\/6\\s+((?:${fiveSquaresRegex}\\s*){1,6})`, "u");
  }

  serializeResult(gameResult: RegExpMatchArray) {
    const [, gameNumber, numGuessesStr, guessesBlock] = gameResult;
    const isWin = numGuessesStr !== "X";

    // First encode the game number amd num guesses.
    const numGuesses = isWin ? parseInt(numGuessesStr) : 6;
    let encoded = intToBase64(parseInt(gameNumber) * 7 + (isWin ? numGuesses - 1 : 6));
    const guesses = splitEmojiLines(guessesBlock);
    if (guesses.length !== numGuesses) {
      throw new Error(`Expected ${numGuesses} guesses, got ${guesses.length}`);
    }
    // Each guess is a closeness value (0-10). We can fit three of these into
    // 2 base64 chars, since 11^3 < 64^2.
    const numericRows = guesses.map(emojiSquareScaleToNumber);
    let guessEncoding = 0;
    for (let guessIdx = 0; guessIdx < numericRows.length; ++guessIdx) {
      const closeness = numericRows[guessIdx];
      // Encode three rows at a time.
      const pow = guessIdx % 3;
      guessEncoding += closeness * 11 ** pow;

      // Write the encoding every third row, or at the end.
      if (pow === 2 || guessIdx === guesses.length - 1) {
        encoded += intToBase64(guessEncoding, guessEncodingWidth);
        guessEncoding = 0;
      }
    }
    const score = isWin
      ? GameScore.Win
      : numericRows.some((val) => val >= 9)
        ? GameScore.NearWin
        : GameScore.Loss;
    return { score, serializedResult: encoded };
  }

  deserialize(serializedResult: string) {
    if (
      // Metadata plus one or two three-guess blocks.
      ![
        DEFAULT_WORD_WIDTH + guessEncodingWidth,
        DEFAULT_WORD_WIDTH + 2 * guessEncodingWidth,
      ].includes(serializedResult.length)
    ) {
      throw new Error(`Invalid Tradle result: ${serializedResult}`);
    }
    // Start with the metadata.
    let intResult: number;
    [intResult, serializedResult] = nextInt(serializedResult);
    let numGuesses = intResult % 7;
    const isWin = numGuesses !== 6;
    if (isWin) {
      // Switch back to 1-indexed.
      numGuesses++;
    }
    const gameNumber = Math.floor(intResult / 7);

    // Then decode the guess rows.
    const guesses: string[] = [];
    for (let guessIdx = 0; guessIdx < numGuesses; ++guessIdx) {
      const pow = guessIdx % 3;
      if (pow === 0) {
        [intResult, serializedResult] = nextInt(serializedResult, guessEncodingWidth);
      }
      const closeness = Math.floor(intResult / 11 ** pow) % 11;
      guesses.push(numberToEmojiSquareScale(closeness, emojiSquares.lightGrey));
    }
    return `#Tradle #${gameNumber} ${isWin ? numGuesses : "X"}/6\n${guesses.join("\n")}`;
  }
}
