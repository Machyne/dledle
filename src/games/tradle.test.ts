import { GameScore } from "../baseGame";
import { Tradle } from "./tradle";

const sampleOddRowWin = `
#Tradle #663 5/6
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©â¬œ
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©â¬œ
ðŸŸ©ðŸŸ©ðŸŸ©â¬œâ¬œ
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ¨
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©
`;

const sampleEvenRowWin = `
#Tradle #663 2/6
ðŸŸ©ðŸŸ©ðŸŸ©â¬œâ¬œ
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©
`;

const sampleNearWin = `
#Tradle #663 X/6
ðŸŸ©ðŸŸ©ðŸŸ©â¬œâ¬œ
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ¨â¬œ
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ¨
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ¨
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©â¬œ
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ¨
`;

const sampleLoss = `
#Tradle #663 X/6
ðŸŸ©ðŸŸ©ðŸŸ©â¬œâ¬œ
ðŸŸ©ðŸŸ©ðŸŸ©â¬œâ¬œ
ðŸŸ©ðŸŸ©ðŸŸ¨â¬œâ¬œ
ðŸŸ©ðŸŸ©ðŸŸ©â¬œâ¬œ
ðŸŸ©ðŸŸ©ðŸŸ©ðŸŸ¨â¬œ
ðŸŸ©ðŸŸ©ðŸŸ©â¬œâ¬œ
`;

describe("Tradle", () => {
  const tradle = new Tradle();

  it("can parse with extra link text", () => {
    const match = tradle.resultRegex.exec(sampleOddRowWin + "\nhttps://oec.world/en/tradle");
    expect(match).not.toBeNull();
    const { score, serializedResult } = tradle.serializeResult(match!);
    expect(score).toEqual(GameScore.Win);
    expect(tradle.deserialize(serializedResult)).toEqual(sampleOddRowWin.trim());
  });

  for (const [name, input, expectedScore] of [
    ["sampleEvenRowWin", sampleEvenRowWin, GameScore.Win],
    ["sampleOddRowWin", sampleOddRowWin, GameScore.Win],
    ["sampleNearWin", sampleNearWin, GameScore.NearWin],
    ["sampleLoss", sampleLoss, GameScore.Loss],
  ] as Array<[string, string, GameScore]>) {
    it(`should serialize and deserialize ${name} correctly`, () => {
      const match = tradle.resultRegex.exec(input);
      expect(match).not.toBeNull();
      const { score, serializedResult } = tradle.serializeResult(match!);

      expect(score).toEqual(expectedScore);
      expect(tradle.deserialize(serializedResult)).toEqual(input.trim());
    });
  }
});
