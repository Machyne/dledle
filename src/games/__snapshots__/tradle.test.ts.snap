// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tradle serialized data should match snapshot 1`] = `
{
  "sampleEvenRowWin": {
    "score": "Win",
    "serializedResult": "iIB0B",
  },
  "sampleEvenRowWin trimmed": {
    "score": "Win",
    "serializedResult": "iIB0B",
  },
  "sampleEvenRowWin with extra link text": {
    "score": "Win",
    "serializedResult": "iIB0B",
  },
  "sampleLoss": {
    "score": "Loss",
    "serializedResult": "nIBlKpM",
  },
  "sampleNearWin": {
    "score": "NearWin",
    "serializedResult": "nIBUSiS",
  },
  "sampleOddRowWin": {
    "score": "Win",
    "serializedResult": "lIB2M3B",
  },
}
`;
